overlord/snapstate/snapstate.go

```diff
@@ -471,6 +473,19 @@ func WaitRestart(task *state.Task, snapsup *SnapSetup) (err error) {
 		return fmt.Errorf("there was a snapd rollback across the restart")
 	}
 
+	// if we have restarted and snapd was refreshed, then we need to generate
+	// snapd wrappers again with current snapd, as the logic of generating
+	// wrappers may have changed between previous and new snapd code.
+	if snapsup.Type == snap.TypeSnapd && !release.OnClassic {
+		snapdInfo, err := snap.ReadCurrentInfo(snapsup.SnapName())
+		if err != nil {
+			return fmt.Errorf("cannot get current snapd snap info: %v", err)
+		}
+		if err := generateSnapdWrappers(snapdInfo); err != nil {
```

Comment:
Those will not be active until we reload systemd and snapd. Is that a problem?

Reply:
This calls wrappers.AddSnapdSnapServices(), and daemon-reload is done there.

Reply:
Should we restart snapd as well, if there were actual changes?

Reply:
Good question, but I think the answer is no, because:

- we are running this code in a new snapd already (after a successfull restart).
- if the changes we introduce in wrappers actually require another snapd restart, then this should be introduced with such changes (they are guaranteed to be run here, in the new snapd).

I hope this makes sense, perhaps @mvo5 or @pedronis should verify my thinking.

Reply:
the main reason to want to restart snapd would be if the .service files have changed, but that should be already taken care correctly but the original generateSnapdWrappers invocation, that one has code to deal with the systemd units taking them from the new revision already. So I agree with @stolowski there's is no reasons atm to restart snapd again from he, if we needed to, we would have likely to keep some state and end up here again but knowing that we can progress. We might want to leave a TODO about future cases that might need that maybe, hopefully not anytime soon.